transport/azure/azure.go

```diff
@@ -32,51 +44,119 @@ func New() (http.RoundTripper, error) {
-	containerURL := t.serviceURL.NewContainerURL(strings.ToLower(req.URL.Host))
-	blobURL := containerURL.NewBlockBlobURL(strings.TrimPrefix(req.URL.Path, "/"))
+		client, err = azblob.NewClient(endpointURL.String(), defaultAzureCredential, nil)
+	}
 
-	start, end, err := httprange.Parse(req.Header.Get("Range"))
 	if err != nil {
-		return httprange.InvalidHTTPRangeResponse(req), nil
+		return nil, err
 	}
 
-	length := end - start + 1
-	if end <= 0 {
-		length = azblob.CountToEnd
+	return transport{client}, nil
+}
+
+func (t transport) RoundTrip(req *http.Request) (*http.Response, error) {
+	container := req.URL.Host
+	key := req.URL.Path
+
+	var statusCode = http.StatusOK
+
+	header := make(http.Header)
+	opts := &blob.DownloadStreamOptions{}
+
+	if r := req.Header.Get("Range"); len(r) != 0 {
+		start, end, err := httprange.Parse(r)
+		if err != nil {
+			return httprange.InvalidHTTPRangeResponse(req), err
+		}
+
+		if end != 0 {
+			length := end - start + 1
+			if end < 0 {
+				length = -1
+			}
+
+			opts.Range = blob.HTTPRange{
+				Offset: start,
+				Count:  length,
+			}
+		}
+
+		statusCode = http.StatusPartialContent
 	}
 
-	get, err := blobURL.Download(req.Context(), start, length, azblob.BlobAccessConditions{}, false, azblob.ClientProvidedKeyOptions{})
+	result, err := t.client.DownloadStream(req.Context(), container, strings.TrimPrefix(key, "/"), opts)
 	if err != nil {
-		return nil, err
+		if azError, ok := err.(*azcore.ResponseError); !ok || azError.StatusCode < 100 || azError.StatusCode == 301 {
+			return nil, err
+		} else {
+			body := strings.NewReader(azError.Error())
+			return &http.Response{
+				StatusCode:    azError.StatusCode,
+				Proto:         "HTTP/1.0",
+				ProtoMajor:    1,
+				ProtoMinor:    0,
+				Header:        header,
+				ContentLength: int64(body.Len()),
+				Body:          io.NopCloser(body),
+				Close:         false,
+				Request:       req,
+			}, err
+		}
 	}
 
-	if config.ETagEnabled && start == 0 && end == azblob.CountToEnd {
-		etag := string(get.ETag())
-
-		if etag == req.Header.Get("If-None-Match") {
-			if body := get.Response().Body; body != nil {
-				get.Response().Body.Close()
-			}
+	if config.ETagEnabled && result.ETag != nil {
+		azETag := string(*result.ETag)
+		header.Set("ETag", azETag)
 
+		if etag := req.Header.Get("If-None-Match"); len(etag) > 0 && azETag == etag {
 			return &http.Response{
 				StatusCode:    http.StatusNotModified,
 				Proto:         "HTTP/1.0",
 				ProtoMajor:    1,
 				ProtoMinor:    0,
-				Header:        make(http.Header),
+				Header:        header,
 				ContentLength: 0,
 				Body:          nil,
 				Close:         false,
 				Request:       req,
-			}, nil
+			}, err
 		}
 	}
 
-	return get.Response(), nil
+	header.Set("Accept-Ranges", "bytes")
+	header.Set("Content-Length", strconv.FormatInt(*result.ContentLength, 10))
+	header.Set("Content-Type", *result.ContentType)
```

Comment:
I'd check that `result.ContentLength` and `result.ContentType` aren't `nil`. it's better safe than sorry :)